File: packages/core/src/session-types.ts
// Session type vocabulary — extracted from session.ts (file-length limit).
// session.ts re-exports this surface verbatim; import from there externally.
import {
  type AlwaysAllowPaths,
  type AskPermissionRequest,
  type MessageToolPermission,
  type UserToolPermission,
} from "./common/permissions";
import { type CreateOpenAIClient, type CreateSecondaryClient } from "./tools/executor";
import { type ToolDefinition } from "./prompt";
import type { McpServerConfig, PermissionScope, PermissionSettings, RoutingSettings } from "./settings";
import type { SandboxBackendStatus } from "./sandbox/backend/interface";
import type { WebPageFetcher } from "./common/tool-types";

/** Memory provider interface — implemented by @deeporca/memory or any compatible provider. */
export interface MemoryProvider {
  recall(
    query: string,
    sessionKey: string
  ): Promise<{ prependContext?: string; appendSystemContext?: string; recallStrategy?: string } | null>;
  capture(turn: {
    userText: string;
    assistantText: string;
    sessionKey: string;
    sessionId?: string;
    /** Last user + assistant messages, so the provider can persist them to L0.
     * Each entry: { role: "user"|"assistant"|"system", content: string, id?: string, timestamp?: number(epoch-ms) }.
     * "system" entries are internal hints (task lineage / recall hints) — kept
     * in L0 for search, excluded from L1 extraction (Phase 4 / T4.3).
     * Provider falls back to synthesizing two messages from userText/assistantText when omitted. */
    messages?: Array<{ role: "user" | "assistant" | "system"; content: string; id?: string; timestamp?: number }>;
  }): Promise<unknown>;
  searchMemories(query: string, limit?: number): Promise<{ text: string; total: number } | null>;
  /**
   * Agent-callable read-only retrieval tools contributed to the LLM tool
   * surface (tdai_memory_search / tdai_conversation_search — Phase 4 / T4.1,
   * specs/memory-remediation). Optional so minimal providers keep working.
   */
  getToolDefinitions?(): ToolDefinition[];
  /** Execute one tool call from {@link getToolDefinitions}. Throws on misuse. */
  executeTool?(name: string, args: Record<string, unknown>): Promise<string>;
  isAvailable(): boolean;
}

export type ChatCompletionDebugOptions = {
  enabled?: boolean;
  location: string;
  baseURL?: string;
  params?: Record<string, unknown>;
};

export type SessionStatus =
  | "failed"
  | "pending"
  | "processing"
  | "waiting_for_user"
  | "completed"
  | "interrupted"
  | "paused"
  | "ask_permission"
  | "permission_denied";

export type ModelUsage = {
  prompt_tokens: number;
  completion_tokens: number;
  total_tokens: number;
  completion_tokens_details?: Record<string, unknown>;
  prompt_tokens_details?: Record<string, unknown>;
  prompt_cache_hit_tokens?: number;
  prompt_cache_miss_tokens?: number;
  total_reqs?: number;
};

export type SessionProcessEntry = {
  startTime: string;
  command: string;
  timeoutMs?: number;
  deadlineAt?: string;
  timedOut?: boolean;
};

export type BashTimeoutAdjustment = {
  processId: string;
  timeoutMs: number;
  deadlineAt: string;
  timedOut: boolean;
};

export type SessionEntry = {
  id: string;
  summary: string | null;
  assistantReply: string | null;
  assistantThinking: string | null;
  assistantRefusal: string | null;
  toolCalls: unknown[] | null;
  status: SessionStatus;
  failReason: string | null;
  usage: ModelUsage | null;
  usagePerModel: Record<string, ModelUsage> | null;
  activeTokens: number;
  createTime: string;
  updateTime: string;
  processes: Map<string, SessionProcessEntry> | null; // {pid: process info}
  askPermissions?: AskPermissionRequest[];
  planMode?: boolean;
  /** Task trajectory binding (specs/task-tree P1): reverse pointer to the branch this session executes. */
  taskRef?: { treeId: string; branch: string; nodeId: string };
  /**
   * True for silent-subagent sessions (specs/index-knowledge-rework T2):
   * excluded from the session list and their messages never surface in the
   * user's conversation view — background pipelines (index.build-all
   * arch-scan) use these so workspace building leaks zero chat behavior.
   */
  isSilentSubagent?: boolean;
};

export type SessionsIndex = {
  version: 1;
  entries: SessionEntry[];
  originalPath: string;
};

export type SessionMessageRole = "system" | "user" | "assistant" | "tool";

export type MessageMeta = {
  function?: unknown;
  paramsMd?: string;
  resultMd?: string;
  asThinking?: boolean;
  isSummary?: boolean;
  isModelChange?: boolean;
  skill?: SkillInfo;
  permissions?: MessageToolPermission[];
  userPrompt?: UserPromptContent;
};

export type SessionMessage = {
  id: string;
  sessionId: string;
  role: SessionMessageRole;
  content: string | null;
  contentParams: unknown | null;
  messageParams: unknown | null;
  compacted: boolean;
  visible: boolean;
  createTime: string;
  updateTime: string;
  meta?: MessageMeta;
  html?: string;
  checkpointHash?: string;
};

export type UndoTarget = {
  message: SessionMessage;
  index: number;
  canRestoreCode: boolean;
};

export type UserPromptContent = {
  text?: string;
  imageUrls?: string[];
  skills?: SkillInfo[];
  permissions?: UserToolPermission[];
  alwaysAllows?: PermissionScope[];
  /** Path-level "always allow" (task 14): narrow grants, persisted per-path. */
  alwaysAllowPaths?: AlwaysAllowPaths;
  planMode?: boolean;
};

export type SkillInfo = {
  name: string;
  path: string;
  description: string;
  isLoaded?: boolean;
  allowImplicitInvocation?: boolean;
  /** True when this skill lives inside a plugin package (hidden from Skills tab). */
  pluginOwned?: boolean;
  /**
   * Optional compositional-routing metadata (R2 contract). All may be absent —
   * skills without them route exactly as before. Consumed by G3's Compose stage
   * (ioTypeCoercion / categoryJaccard in routing/composer.ts).
   */
  categories?: string[];
  /** Input types this skill consumes (e.g. ["markdown", "file-list"]). */
  inputs?: string[];
  /** Output types this skill produces (e.g. ["html", "pdf"]). */
  outputs?: string[];
};

/**
 * Orca built-in plugin descriptor. Built-in plugins are a first-class extension
 * type parallel to Skills and MCP servers. They ship inside the core package
 * (`templates/plugins/<name>/`) and can never be uninstalled or disabled.
 *
 * Plugin directory layout:
 *   templates/plugins/<name>/plugin.json   – manifest (name, version, category…)
 *   templates/plugins/<name>/PLUGIN.md     – instruction document injected into prompts
 */
export type BuiltinPluginInfo = {
  name: string;
  version: string;
  description: string;
  category: string;
  /** Built-in plugins are never removable. */
  removable: false;
  /** Display path, e.g. "builtin-plugin:browser-skill". */
  path: string;
};

/**
 * A built-in plugin group — the user-facing "plugin card" in the plugin center.
 * Groups unify related skills, MCP servers, and plugin descriptors that belong
 * to the same tool/service into a single displayable unit. The manifest lives
 * at `templates/builtin-plugins.json` (shipped with the product, read-only);
 * this type carries the manifest plus the *resolved* members matched at call
 * time against the live skill/MCP/plugin lists.
 */
export type BuiltinPluginGroup = {
  id: string;
  name: string;
  description: string;
  category: string;
  /** Optional icon identifier (e.g. "flutter"); the renderer maps it to an icon. */
  icon?: string;
  /** When set, the group only loads on the matching OS. The renderer shows it
   * greyed-out on other platforms so users know the capability exists. */
  platform?: "darwin" | "linux";
  /** Skills matched into this group (from the bundled + user skill lists). */
  skills: SkillInfo[];
  /** MCP servers matched into this group (by name or `prefix:*` glob). */
  mcpServers: McpServerConfigEntry[];
  /** Built-in plugin descriptors matched into this group (by name). */
  plugins: BuiltinPluginInfo[];
  /** Actions matched into this group (by id prefix, e.g. "review.*" or "browser.*"). */
  actions: Array<{ id: string; description: string }>;
};

/** Minimal MCP server shape needed for group resolution (name + enabled). */
export type McpServerConfigEntry = {
  name: string;
  config: McpServerConfig;
  builtin?: boolean;
  /** Whether the server is enabled (not in the disabled set). For display in group cards. */
  enabled?: boolean;
  /** Connection status summary (ready/failed/connecting), if known. */
  status?: string;
};

/** Raw manifest entry in `builtin-plugins.json` (before resolution). */
export type BuiltinPluginGroupManifest = {
  id: string;
  name: string;
  description: string;
  category: string;
  icon?: string;
  platform?: "darwin" | "linux";
  skills?: string[];
  mcp?: string[];
  plugins?: string[];
  actions?: string[];
};

/**
 * The settings slice SessionManager consumes (host-injected, narrow structural
 * type — desktop passes the full resolveCurrentSettings result; tests pass the
 * minimum their fakes need).
 */
export type SessionResolvedSettings = {
  model: string;
  /** Configured endpoints (primary-family + lightweight availability checks). */
  endpoints?: ReadonlyArray<{
    id?: string;
    baseURL?: string;
    apiKey?: string;
    models?: ReadonlyArray<{ id: string }>;
  }>;
  primaryEndpointId?: string;
  /** Explicit user-configured secondary model + resolved credentials. */
  secondaryModel?: string;
  secondaryApiKey?: string;
  secondaryBaseURL?: string;
  /** User override for the compaction trigger (tokens); undefined = registry default. */
  compactTokenThreshold?: number;
  webSearchTool?: string;
  mcpServers?: Record<string, McpServerConfig>;
  permissions?: Required<PermissionSettings>;
  workspaceTrust?: "trusted" | "quarantine";
  enabledSkills?: Record<string, boolean>;
  routing?: RoutingSettings;
  visionModel?: string;
  visionApiKey?: string;
  streamIdleTimeoutMs?: number;
};

export type SessionManagerOptions = {
  projectRoot: string;
  createOpenAIClient: CreateOpenAIClient;
  /**
   * Secondary-model client factory — tier-2 fallback for background LLM tasks
   * when the primary family has no lightweight model. Defaults to the real
   * factory from openai-client; injectable for tests.
   */
  createSecondaryClient?: CreateSecondaryClient;
  /** Host-injected rendered-page fetcher for the built-in WebFetch tool. */
  fetchWebPage?: WebPageFetcher;
  getResolvedSettings: () => SessionResolvedSettings;
  renderMarkdown: (text: string) => string;
  onAssistantMessage: (message: SessionMessage, shouldConnect: boolean) => void;
  onSessionEntryUpdated?: (entry: SessionEntry) => void;
  onLlmStreamProgress?: (progress: LlmStreamProgress) => void;
  /** Behavioral-memory provider (activity-frames pipeline B, host-injected). Returns a compact context block or null. */
  buildBehaviorContext?: () => string | null;
  onMcpStatusChanged?: () => void;
  /** Sandbox backend selection outcome per session (active or degraded). */
  onSandboxStatusChanged?: (status: SandboxBackendStatus) => void;
  onProcessStdout?: (pid: number, chunk: string) => void;
};

export type LlmStreamProgress = {
  requestId: string;
  sessionId?: string;
  startedAt: string;
  estimatedTokens: number;
  formattedTokens: string;
  phase: "start" | "update" | "end";
};

File: deeporca-win/src/DeepOrca.Core/Types/SessionTypes.cs
using System.Text.Json.Nodes;

namespace DeepOrca.Core.Types;

// SessionTypes — 会话核心数据模型（对拍 apple SessionTypes.swift / 上游 session-types.ts）
// 存储 camelCase（CoreJson source-gen），与 TS / apple JSONL 字节兼容。

public enum SessionStatus
{
    Failed,
    Pending,
    Processing,
    WaitingForUser,
    Completed,
    Interrupted,
    Paused,
    AskPermission,
    PermissionDenied,
}

public enum SessionMessageRole
{
    System,
    User,
    Assistant,
    Tool,
}

public static class SessionWire
{
    /// <summary>存储/索引 wire 名（snake_case，如 "waiting_for_user"）↔ enum。</summary>
    public static string Wire(this SessionStatus status) => status switch
    {
        SessionStatus.Failed => "failed",
        SessionStatus.Pending => "pending",
        SessionStatus.Processing => "processing",
        SessionStatus.WaitingForUser => "waiting_for_user",
        SessionStatus.Completed => "completed",
        SessionStatus.Interrupted => "interrupted",
        SessionStatus.Paused => "paused",
        SessionStatus.AskPermission => "ask_permission",
        SessionStatus.PermissionDenied => "permission_denied",
        _ => "pending",
    };

    public static SessionStatus StatusFromWire(string name) => name switch
    {
        "failed" => SessionStatus.Failed,
        "pending" => SessionStatus.Pending,
        "processing" => SessionStatus.Processing,
        "waiting_for_user" => SessionStatus.WaitingForUser,
        "completed" => SessionStatus.Completed,
        "interrupted" => SessionStatus.Interrupted,
        "paused" => SessionStatus.Paused,
        "ask_permission" => SessionStatus.AskPermission,
        "permission_denied" => SessionStatus.PermissionDenied,
        _ => SessionStatus.Pending,
    };

    public static string Wire(this SessionMessageRole role) => role switch
    {
        SessionMessageRole.System => "system",
        SessionMessageRole.User => "user",
        SessionMessageRole.Assistant => "assistant",
        SessionMessageRole.Tool => "tool",
        _ => "user",
    };

    public static SessionMessageRole RoleFromWire(string name) => name switch
    {
        "system" => SessionMessageRole.System,
        "user" => SessionMessageRole.User,
        "assistant" => SessionMessageRole.Assistant,
        "tool" => SessionMessageRole.Tool,
        _ => SessionMessageRole.User,
    };
}

public sealed record ModelUsage
{
    public int PromptTokens { get; init; }
    public int CompletionTokens { get; init; }
    public int TotalTokens { get; init; }
    public Dictionary<string, int>? CompletionTokensDetails { get; init; }
    public Dictionary<string, int>? PromptTokensDetails { get; init; }
    public int? PromptCacheHitTokens { get; init; }
    public int? PromptCacheMissTokens { get; init; }
    public int? TotalReqs { get; init; }

    public static ModelUsage Empty { get; } = new();

    public void Deconstruct(out int prompt, out int completion, out int total) =>
        (prompt, completion, total) = (PromptTokens, CompletionTokens, TotalTokens);

    /// <summary>逐轮累加（record init-only：返回合并结果，调用方持有新值）。</summary>
    public ModelUsage Add(ModelUsage other) => new()
    {
        PromptTokens = PromptTokens + other.PromptTokens,
        CompletionTokens = CompletionTokens + other.CompletionTokens,
        TotalTokens = TotalTokens + other.TotalTokens,
        PromptCacheHitTokens = (PromptCacheHitTokens ?? 0) + (other.PromptCacheHitTokens ?? 0),
        PromptCacheMissTokens = (PromptCacheMissTokens ?? 0) + (other.PromptCacheMissTokens ?? 0),
        TotalReqs = (TotalReqs ?? 0) + (other.TotalReqs ?? 1),
    };
}

public sealed record SessionProcessEntry
{
    public required string StartTime { get; init; }
    public required string Command { get; init; }
    public int? TimeoutMs { get; init; }
    public string? DeadlineAt { get; init; }
    public bool? TimedOut { get; init; }
}

public sealed record TaskRef
{
    public required string TreeId { get; init; }
    public required string Branch { get; init; }
    public required string NodeId { get; init; }
}

public sealed record SkillInfo
{
    public required string Name { get; init; }
    public required string Path { get; init; }
    public required string Description { get; init; }
    public bool? IsLoaded { get; init; }
    public bool? AllowImplicitInvocation { get; init; }
    public bool? PluginOwned { get; init; }
    public List<string>? Categories { get; init; }
    public List<string>? Inputs { get; init; }
    public List<string>? Outputs { get; init; }
}

public sealed record UserPromptContent
{
    public string? Text { get; init; }
    public List<string>? ImageUrls { get; init; }
    public List<SkillInfo>? Skills { get; init; }
    public List<UserToolPermission>? Permissions { get; init; }
    public List<PermissionScope>? AlwaysAllows { get; init; }
    public AlwaysAllowPaths? AlwaysAllowPaths { get; init; }
    public bool? PlanMode { get; init; }
}

public sealed record SessionEntry
{
    public required string Id { get; init; }
    public string? Summary { get; set; }
    public string? AssistantReply { get; set; }
    public string? AssistantThinking { get; set; }
    public string? AssistantRefusal { get; set; }
    public JsonNode? ToolCalls { get; set; }
    public SessionStatus Status { get; set; } = SessionStatus.Pending;
    public string? FailReason { get; set; }
    public ModelUsage? Usage { get; set; }
    /// <summary>Keyed by model id（usagePerModel 统计，M5 TokenSummary 消费）。</summary>
    public Dictionary<string, ModelUsage>? UsagePerModel { get; set; }
    public int ActiveTokens { get; set; }
    public string CreateTime { get; init; } = DateTimeOffset.UtcNow.ToString("o");
    public string UpdateTime { get; set; } = DateTimeOffset.UtcNow.ToString("o");
    /// <summary>
    /// 后台进程跟踪（key = pid）。注意 AGENTS.md 会话索引不变量：内存形态与磁盘形态的
    /// 归一化边界由 SessionStore 负责，此 record 只承载磁盘形状。
    /// </summary>
    public Dictionary<string, SessionProcessEntry>? Processes { get; set; }
    public List<AskPermissionRequest>? AskPermissions { get; set; }
    public bool? PlanMode { get; set; }
    public TaskRef? TaskRef { get; set; }
    public bool? IsSilentSubagent { get; set; }
    public string? WorkspaceDir { get; set; }
}

public sealed record MessageMeta
{
    /// <summary>assistant tool_calls 镜像（apple 形状：[{id,name,arguments}]）。</summary>
    public JsonNode? Function { get; init; }
    public string? ParamsMd { get; init; }
    public string? ResultMd { get; init; }
    public bool? AsThinking { get; init; }
    public bool? IsSummary { get; init; }
    public bool? IsModelChange { get; init; }
    public SkillInfo? Skill { get; init; }
    public List<MessageToolPermission>? Permissions { get; init; }
    public UserPromptContent? UserPrompt { get; init; }
}

public sealed record SessionMessage
{
    public string Id { get; init; } = Guid.NewGuid().ToString();
    public required string SessionId { get; init; }
    public required SessionMessageRole Role { get; init; }
    public string? Content { get; init; }
    /// <summary>user/system 多模态附加 parts（TS contentParams 形状）。</summary>
    public JsonNode? ContentParams { get; init; }
    /// <summary>
    /// 线上协议扩展（TS messageParams 形状）：assistant 持 tool_calls；tool 持 tool_call_id
    /// —— 消息转换器按 id 配对（M1.4），存储原样透传。
    /// </summary>
    public JsonNode? MessageParams { get; init; }
    public bool Compacted { get; init; }
    public bool Visible { get; init; } = true;
    public string CreateTime { get; init; } = DateTimeOffset.UtcNow.ToString("o");
    public string UpdateTime { get; init; } = DateTimeOffset.UtcNow.ToString("o");
    public MessageMeta? Meta { get; init; }
    public string? Html { get; init; }
    public string? CheckpointHash { get; init; }

    public static SessionMessage Create(
        string sessionId,
        SessionMessageRole role,
        string? content,
        JsonNode? messageParams = null,
        MessageMeta? meta = null,
        bool compacted = false,
        JsonNode? contentParams = null) =>
        new()
        {
            SessionId = sessionId,
            Role = role,
            Content = content,
            MessageParams = messageParams,
            Meta = meta,
            Compacted = compacted,
            ContentParams = contentParams,
        };
}

public sealed record SessionsIndex
{
    public int Version { get; init; } = 1;
    public List<SessionEntry> Entries { get; set; } = [];
    public string? OriginalPath { get; init; }
}

public enum StreamPhase
{
    Start,
    Update,
    End,
}

public sealed record LlmStreamProgress
{
    public required string RequestId { get; init; }
    public string? SessionId { get; init; }
    public string StartedAt { get; init; } = DateTimeOffset.UtcNow.ToString("o");
    public int? EstimatedTokens { get; init; }
    public string? FormattedTokens { get; init; }
    public required StreamPhase Phase { get; init; }
}
